### DISCUSSION TITLE
Insert client's IP address into an Autoinstallation Variable

### DISCUSSION THREAD

**Discussion by kurzandras** [uyuni-project/uyuni#9158] [General]
Hi all,

Just a quick question: I would like to upgrade a client from sles12sp5 to sles15sp3 and I noticed that during the autoinstallation the client tries to get its IP address from DHCP. I do not want this to happen, I would like to use the same IP what the client had before. Is there a simple solution for this? I thought that I would add an IP address variable to the Autoinstallation Variables which I could reference from the autoyast profile using bootproto static, but im not sure how to get the client's IP address from the UI dynamically. Is there a solution for this? Or any other solutions for this issue?

Thank you in advance!

**Comment by aaannz** [comment]
Hi,

check [Linuxrc](https://en.opensuse.org/SDB:Linuxrc#Network_Config). This needs to be added as a kernel command line options for the autoinstallation profile.

**Reply by kurzandras** [reply]
Hi @aaannz ! Thank you very much for your reply. I have taken a look on the page you linked, but I can not see how exactly this is helpful in this particular case. 

My main goal is just to preserve the original IP address of the client machine, so it would be fine if I could use the same network config as before, or there would be no network reconfiguration during the upgrade process at all.

Could you please provide just a little more guidance? I really appreciate your help!

**Reply by aaannz** [reply]
I don't think this is implemented. For autoinstallation you need to have network so that autoyast profile and package sources can be fetched from the uyuni server. This installation network is setup by linuxrc and I do not think linuxrc is able to mount existing volumes to read it's configuration.

What network settings will be used after upgrade, that is specified by [keep_install_network](https://doc.opensuse.org/projects/autoyast/#CreateProfile-Network-Resource) autoyast option, which is by default `true` so original network configuration should be reused. However this part is unrelated to network settings used by installer.

**Reply by kurzandras** [reply]
Thank you for your answer! Understood, its clear. The only question now is how I am able to get the Uyuni clients IP address (perhaps from the Uyuni UI through a special variable) and put it into the profile.

Meanwhile I have tried to experiment with the upgrade and I have tried the following profile with only a hardcoded IP address, but it seems that during upgrade it still tried to turn to DHCP despite the bootproto=static option:

```shell
<?xml version="1.0"?>
<!DOCTYPE profile>
<profile xmlns="http://www.suse.com/1.0/yast2ns" xmlns:config="http://www.suse.com/1.0/configns">
    <general>
    $SNIPPET('spacewalk/sles_no_signature_checks')
        <mode>
            <confirm config:type="boolean">false</confirm>
        </mode>
    </general>
    <add-on>
        <add_on_products config:type="list">
            <listentry>
                <ask_on_error config:type="boolean">true</ask_on_error>
                <media_url>http://$redhat_management_server/ks/dist/child/sle-product-sles15-sp3-updates-x86_64/$distrotree</media_url>
                <name>SLE-Product-SLES15-SP3-Updates for x86_64</name>
                <product>SUSE Linux Enterprise Server 15 SP3 x86_64</product>
                <product_dir>/</product_dir>
            </listentry>
            <listentry>
                <ask_on_error config:type="boolean">true</ask_on_error>
                <media_url>http://$redhat_management_server/ks/dist/child/sle-manager-tools15-pool-x86_64-sp3/$distrotree</media_url>
                <name>SLE-Manager-Tools15-Pool for x86_64 SP3</name>
                <product>SUSE Manager Tools 15 x86_64</product>
                <product_dir>/</product_dir>
            </listentry>
            <listentry>
                <ask_on_error config:type="boolean">true</ask_on_error>
                <media_url>http://$redhat_management_server/ks/dist/child/sle-manager-tools15-updates-x86_64-sp3/$distrotree</media_url>
                <name>SLE-Manager-Tools15-Updates for x86_64 SP3</name>
                <product>SUSE Manager Tools 15 x86_64</product>
                <product_dir>/</product_dir>
            </listentry>
            <listentry>
                <ask_on_error config:type="boolean">true</ask_on_error>
                <media_url>http://$redhat_management_server/ks/dist/child/sle-module-basesystem15-sp3-pool-x86_64/$distrotree</media_url>
                <name>SLE-Module-Basesystem15-SP3-Pool for x86_64</name>
                <product>Basesystem Module 15 SP3 x86_64</product>
                <product_dir>/</product_dir>
            </listentry>
            <listentry>
                <ask_on_error config:type="boolean">true</ask_on_error>
                <media_url>http://$redhat_management_server/ks/dist/child/sle-module-basesystem15-sp3-updates-x86_64/$distrotree</media_url>
                <name>SLE-Module-Basesystem15-SP3-Updates for x86_64</name>
                <product>Basesystem Module 15 SP3 x86_64</product>
                <product_dir>/</product_dir>
            </listentry>
            <listentry>
                <ask_on_error config:type="boolean">true</ask_on_error>
                <media_url>http://$redhat_management_server/ks/dist/child/sle-module-server-applications15-sp3-pool-x86_64/$distrotree</media_url>
                <name>SLE-Module-Server-Applications15-SP3-Pool for x86_64</name>
                <product>Server Applications Module 15 SP3 x86_64</product>
                <product_dir>/</product_dir>
            </listentry>
            <listentry>
                <ask_on_error config:type="boolean">true</ask_on_error>
                <media_url>http://$redhat_management_server/ks/dist/child/sle-module-server-applications15-sp3-updates-x86_64/$distrotree</media_url>
                <name>SLE-Module-Server-Applications15-SP3-Updates for x86_64</name>
                <product>Server Applications Module 15 SP3 x86_64</product>
                <product_dir>/</product_dir>
            </listentry>
        </add_on_products>
    </add-on>
    <networking>
        <keep_install_network config:type="boolean">false</keep_install_network>
        <interfaces config:type="list">
            <interface>
                <device>eth0</device>
                <bootproto>static</bootproto>
                <ipaddr>171.10.111.202</ipaddr>
                <prefixlen>24</prefixlen>
            </interface>
        </interfaces>
    </networking>
    <upgrade>
        <stop_on_solver_conflict config:type="boolean">true</stop_on_solver_conflict>
    </upgrade>
    <suse_register>
        <do_registration config:type="boolean">false</do_registration>
    </suse_register>
    <backup>
        <sysconfig config:type="boolean">true</sysconfig>
        <modified config:type="boolean">true</modified>
        <remove_old config:type="boolean">false</remove_old>
    </backup>
    <software>
        <products config:type="list">
            <product>SLES</product>
        </products>
    </software>
    <scripts>
        <!-- This script is designed to ensure the certificate is present for any 'https' repositories -->
        <pre-scripts config:type="list">
            <script>
                <rerun config:type="boolean">false</rerun>
                <debug config:type="boolean">false</debug>
                <feedback config:type="boolean">false</feedback>
                <filename>01_import_cert</filename>
                <interpreter>shell</interpreter>
                <location/>
                <notification>Importing SUMA CA certificate</notification>
                <source><![CDATA[
/usr/bin/wget -nv -r -nd --no-check-certificate https://$redhat_management_server/pub/RHN-ORG-TRUSTED-SSL-CERT
mv RHN-ORG-TRUSTED-SSL-CERT /etc/pki/trust/anchors/
/usr/sbin/update-ca-certificates
     ]]>        </source>
            </script>
        </pre-scripts>
        <init-scripts config:type="list">
        $SNIPPET('spacewalk/minion_script')
        </init-scripts>
    </scripts>
</profile>
``` 



Could you please give me some guidance?
Thank you very much in advance!

**Reply by aaannz** [reply]
The installer asking DHCP for IP address even when hardcoded in profile is expected. Because first you need some network setup to be able to download the autoyast profile, only after profile is downloaded, installer can configure the network.
So look at autoyast profile as how system should look like once installed. Not how the system should look like right away from boot.

To setup installer network you need to use Linuxrc kernel options. So in your case something like `ifcfg=eth0=171.10.111.202/24,171.10.111.1` assuming 171.10.111.1 is the gateway/router ip. And you will need to add this to the autoinstallation profile in the uyuni web ui here:
![image](https://github.com/user-attachments/assets/20ca64b3-d45f-4d07-b842-e995385cbca2)

**Reply by kurzandras** [reply]
Thank you very much for your great explanation! Do you think there is a way to use variables from Uyuni in the kernel options? I would need to somehow upgrade many hosts, but if the only option is to hardcode ip addresses into the profile kernel options then i have no idea how to achieve this. How would you solve this problem? Could that work if I add custom variables to the clients for the ip addresses and then i reference them from the kernel options somehow?

**Reply by aaannz** [reply]
Doing this in script is quite a work. Probably more work then setting up a DHCP to temporarily cover these installation parts.

I assume you are doing these system upgrades using provisioning menu and not through PXE, right? Otherwise you would already have DHCP.

Easiest thing would be to edit uyuni salt template [`bootloader/42_uyuni_reinstall.templ`](https://github.com/uyuni-project/uyuni/blob/master/susemanager-utils/susemanager-sls/salt/bootloader/42_uyuni_reinstall.templ). In there you can get client ipv4 from minion grain (`grains.get['ipv4']`), filter out localhost '127.0.0.1' and save if after that `{{ pillar.get ... }}`.

**Reply by kurzandras** [reply]
Thank you very much for your help @aaannz, I was able to achieve it as you wrote. It would really be a great feature in the future if we could use client specific variables from the UI in the kernel options and in the profile. 

One additional question: Now after upgrading a sles12 client to sles15 it seems that it is still subscribed to old channels:
![image](https://github.com/user-attachments/assets/cefd4d58-2eb9-4bbc-a695-51773a6b2fb5)

Do you have any ideas why SLES12 is still there?

**Reply by aaannz** [reply]
First check clients event log in Uyuni webui if package list refresh event finished successfully.

If it is already finished and no error was reported, not sure how it is possible, but the system probably has old SLE12 release package still present there.
Check on the client `zypper se -si release`.

Upgrade should have replaced them with SLE15 SP3 variants however.

**Reply by kurzandras** [reply]
No errors, the package list refresh was successful.

![image](https://github.com/user-attachments/assets/84f75a7b-d9ea-4e74-9a1b-a2bb2e8ef4d5)

What do you suggest in this case?

**Reply by aaannz** [reply]
Before continuing with migrating other systems you will need to adapt that autoyast profile to contain the rest of used modules

- `sle-module-web-scripting15-sp3-updates-x86_64`
- `sle-module-legacy15-sp3-updates-x86_64`
- `sle-module-live-patching15-sp3-updates-x86_64`
- `sle-module-packagehub-subpackages15-sp3-updates-x86_64`

On this system already migrated, I can't say for sure. I would try to alter channels manually and then `zypper dup` on the system.
It is really strange that installer was fine with keeping SLE12 modules without having sles-release for SLE12. I would expected some error message during migration.

**Reply by kurzandras** [reply]
This is only a test system where I am trying to implement the process of an upgrade, so I will revert this host for the next test-run anyway. I did not see any errors during the installation, it was quite smooth. 

Im not sure if the profile is completely fine, I just tried to compose a working initial version:

```xslt
<?xml version="1.0"?>
<!DOCTYPE profile>
<profile xmlns="http://www.suse.com/1.0/yast2ns"
    xmlns:config="http://www.suse.com/1.0/configns">
    <general>
    $SNIPPET('spacewalk/sles_no_signature_checks')
        <mode>
            <confirm config:type="boolean">false</confirm>
        </mode>
    </general>
    <add-on>
      <add_on_products config:type="list">
        <listentry>
          <ask_on_error config:type="boolean">true</ask_on_error>
          <media_url>http://uyuni01.test/ks/dist/child/sle-product-sles15-sp3-updates-x86_64/$distrotree</media_url>
          <name>SLE-Product-SLES15-SP3-Updates for x86_64</name>
          <product>SUSE Linux Enterprise Server 15 SP3 x86_64</product>
          <product_dir>/</product_dir>
        </listentry>
        <listentry>
          <ask_on_error config:type="boolean">true</ask_on_error>
          <media_url>http://uyuni01.test/ks/dist/child/sles15-sp3-uyuni-client-x86_64/$distrotree</media_url>
          <name>Uyuni Client Tools for SLES15 SP3 x86_64</name>
          <product>Uyuni Client Tools 15 x86_64</product>
          <product_dir>/</product_dir>
        </listentry>
        <listentry>
          <ask_on_error config:type="boolean">true</ask_on_error>
          <media_url>http://uyuni01.test/ks/dist/child/sle-module-basesystem15-sp3-pool-x86_64/$distrotree</media_url>
          <name>SLE-Module-Basesystem15-SP3-Pool for x86_64</name>
          <product>Basesystem Module 15 SP3 x86_64</product>
          <product_dir>/</product_dir>
        </listentry>
        <listentry>
          <ask_on_error config:type="boolean">true</ask_on_error>
          <media_url>http://uyuni01.test/ks/dist/child/sle-module-basesystem15-sp3-updates-x86_64/$distrotree</media_url>
          <name>SLE-Module-Basesystem15-SP3-Updates for x86_64</name>
          <product>Basesystem Module 15 SP3 x86_64</product>
          <product_dir>/</product_dir>
        </listentry>
        <listentry>
          <ask_on_error config:type="boolean">true</ask_on_error>
          <media_url>http://uyuni01.test/ks/dist/child/sle-module-server-applications15-sp3-pool-x86_64/$distrotree</media_url>
          <name>SLE-Module-Server-Applications15-SP3-Pool for x86_64</name>
          <product>Server Applications Module 15 SP3 x86_64</product>
          <product_dir>/</product_dir>
        </listentry>
        <listentry>
          <ask_on_error config:type="boolean">true</ask_on_error>
          <media_url>http://uyuni01.test/ks/dist/child/sle-module-server-applications15-sp3-updates-x86_64/$distrotree</media_url>
          <name>SLE-Module-Server-Applications15-SP3-Updates for x86_64</name>
          <product>Server Applications Module 15 SP3 x86_64</product>
          <product_dir>/</product_dir>
        </listentry>
      </add_on_products>
    </add-on>
    <networking>
        <keep_install_network config:type="boolean">true</keep_install_network>
    </networking>
    <upgrade>
        <stop_on_solver_conflict config:type="boolean">true</stop_on_solver_conflict>
    </upgrade>
    <suse_register>
        <do_registration config:type="boolean">false</do_registration>
    </suse_register>
    <backup>
        <sysconfig config:type="boolean">true</sysconfig>
        <modified config:type="boolean">true</modified>
        <remove_old config:type="boolean">false</remove_old>
    </backup>
    <software>
        <install_recommended config:type="boolean">true</install_recommended>
        <products config:type="list">
            <product>SLES</product>
        </products>
    </software>
    <scripts>
        <!-- This script is designed to ensure the certificate is present for any 'https' repositories -->
        <pre-scripts config:type="list">
            <script>
                <rerun config:type="boolean">false</rerun>
                <debug config:type="boolean">false</debug>
                <feedback config:type="boolean">false</feedback>
                <filename>01_import_cert</filename>
                <interpreter>shell</interpreter>
                <location/>
                <notification>Importing SUMA CA certificate</notification>
                <source><![CDATA[
/usr/bin/wget -nv -r -nd --no-check-certificate https://$redhat_management_server/pub/RHN-ORG-TRUSTED-SSL-CERT
mv RHN-ORG-TRUSTED-SSL-CERT /etc/pki/trust/anchors/
/usr/sbin/update-ca-certificates
     ]]>                </source>
            </script>
        </pre-scripts>
        <init-scripts config:type="list">
        $SNIPPET('spacewalk/minion_script')
        </init-scripts>
    </scripts>
</profile>
``` 

Do you see any problems here?

**Reply by kurzandras** [reply]
One additional question: The documentation states that we need to create an activation key for the upgrade process, but it seems that it is not really being used at all. How/where the new activation key is linked to the upgrade process?

**Reply by aaannz** [reply]
I misted your previous question, sorry. The profile looks fine, just is missing those extra channels, so copy
```xml
        <listentry>
          <ask_on_error config:type="boolean">true</ask_on_error>
          <media_url>http://uyuni01.test/ks/dist/child/sle-module-server-applications15-sp3-updates-x86_64/$distrotree</media_url>
          <name>SLE-Module-Server-Applications15-SP3-Updates for x86_64</name>
          <product>Server Applications Module 15 SP3 x86_64</product>
          <product_dir>/</product_dir>
        </listentry>
```
and add those missing channels from my previous comment. Should fix that leftover SLE12 repositories.

About the activation key, it's in the `minion_script` snippet, see https://github.com/uyuni-project/uyuni/blob/master/java/conf/cobbler/snippets/minion_script#L22

**Reply by kurzandras** [reply]
Thank you @aaannz for your answer! I have successfully upgraded a system from SLES12 to SLES15, and afterwards I have installed some patches for SLES15 too. Everything seems to be fine except one thing: Uyuni states that "The system requires a reboot", which does not disappear after reboot. Moreover when I click on "Apply Highstate" the task will get stuck in picked-up state:

![image](https://github.com/user-attachments/assets/9bc73676-3f1a-4226-870d-2308f21a59f3)

Hardware list refresh too:

![image](https://github.com/user-attachments/assets/c862a8d6-e21d-4eb5-a32a-0eae5942e9ea)

Other things like package list update and patching are working fine. I could not see any issues in the server / client logs. How does Uyuni know which system needs to be rebooted? Could you please point me to the right direction regarding troubleshooting?

Thank you very much in advance!
